### PR TITLE
LPD-26659 FEAT Blogs friendly url categories reorder

### PR DESCRIPTION
**Pull request by robertoDiaz**
# What is this trying to solve?

- https://liferay.atlassian.net/browse/LPD-29741 (implementation)
- https://liferay.atlassian.net/browse/LPD-33437 (display context tests)
- https://liferay.atlassian.net/browse/LPD-32762 (migrate playwright test)

Allowing the blog admin to reorder the categories in the friendly URL.
Now the categories must be added in the blog entry to use them in the friendly URL.

# How am I fixing it?

We are using the [Dual List Box](https://clayui.com/docs/components/dual-list-box.html) to allow categories to reorder.

# Screenshots
## Custom friendly URL
image

## Automatic friendly URL
image

# How can you verify that it works?

Otherwise, define a procedure to test the changes using a numbered list. E.g:
1. Go to Content & Data > Blogs > new
2. Add some categories in the categorization section
3. Use these categories in the friendly URL section
4. Publish

You can edit an entry instead of creating a new one.

Also with the existing functional tests (modified in this PR) and some Integration tests also included (run `com.liferay.blogs.web.internal.display.context.test.BlogsEditEntryDisplayContextTest`)

# Playwright test

1. Go to `/modules/test/playwright`
3. Run `npm install` (setup)
4. Run `npm run playwright -- test --project blogs-web --reporter list --grep-invert @LPD-22497` to run the test

## Avoid flakiness by repeating mutiple times

```sh
npm run playwright -- test --project blogs-web --reporter list --grep-invert @LPD-22497 --repeat-each 10
```

### tested on local
image

### CODE REVIEW

**Comment by robertoDiaz** [comment]
cc/ @boton

**Comment by boton** [comment]
ci:test:playwright

**Comment by ambrinchaudhary** [comment]
Frontend part was previously reviewed and approved in https://github.com/liferay-content-management/liferay-portal/pull/5753, it there anything new to review? Otherwise, could we remove the `frontend review needed` label?

**Comment by robertoDiaz** [comment]
> Frontend part was previously reviewed and approved in #5753, it there anything new to review? Otherwise, could we remove the `frontend review needed` label?

Hi @ambrinchaudhary : @boton could provide more information about that (I'm not sure if he pushed more changes from your last review). At least the functional tests have been updated in this new iteration.

**Comment by boton** [comment]
I applied your feedback https://github.com/liferay-content-management/liferay-portal/pull/5753#discussion_r1721350319 here 81f7bc4834571b3aa0b079d42a3c9d8c7b601785 and only left he attribute `hidden` it is enough you are right 😄

**Comment by boton** [comment]
ci:test:sf

**Comment by boton** [comment]
ci:test:playwright

File: modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/display/context/BlogsEditEntryDisplayContext.java [hunk 1]
```diff
@@ -566,7 +567,14 @@ private long[] _getAvailableFriendlyURLAssetCategoryIds() {
...
 
-		_assetCategoryIds = assetEntry.getCategoryIds();
+		List<Long> assetCategoryIdsList = ListUtil.fromArray(
+			assetEntry.getCategoryIds());
+
+		assetCategoryIdsList.removeAll(
+			ListUtil.fromArray(_getCurrentFriendlyURLAssetCategoryIds()));
+
+		_assetCategoryIds = ArrayUtil.toArray(
+			assetCategoryIdsList.toArray(new Long[0]));
```
Comment (adolfopa) [inline modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/display/context/BlogsEditEntryDisplayContext.java]: Why are you mixing `Long[]` and `long[]` here? These are two distinct, non-compatible types.

You should use `ArrayUtil.toLongArray` instead.

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/display/context/BlogsEditEntryDisplayContext.java (hunk 1 above)
Comment (robertoDiaz) [reply modules/apps/blogs/blogs-web/src/main/java/com/liferay/blogs/web/internal/display/context/BlogsEditEntryDisplayContext.java]: @adolfopa and @ambrinchaudhary 

Do you think we could move this forward?

**Comment by boton** [comment]
### Playwright test passing in CI

image

**Review by adolfopa — APPROVED** [review]

**Comment by adolfopa** [comment]
ci:forward

**Comment by robertoDiaz** [comment]
ci:forward

**Pull request closed without merging**